### PR TITLE
Moving tp_class access, and consistent fully-qualified naming for PyPy, to detail::get_tp_name

### PR DESCRIPTION
**Pull request by YannickJadoul**
Increasing consistency, cfr. https://github.com/pybind/pybind11/pull/2349#discussion_r465807968

### CODE REVIEW

File: tests/test_local_bindings.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def test_internal_locals_differ():
     assert m.local_cpp_types_addr() != cm.local_cpp_types_addr()
 
 
-@pytest.mark.xfail("env.PYPY")
+@pytest.mark.xfail("env.PYPY and sys.pypy_version_info < (7, 3, 2)")
```
Comment (bstaletic) [inline tests/test_local_bindings.py:158]: "Expect this to fail on PyPy < 7.3.2"

Didn't we decide to drop older versions already?

Comment (YannickJadoul) [reply tests/test_local_bindings.py:158]: 7.3.2 is not out, yes. And it still fails on 7.3.1 :-)

Comment (YannickJadoul) [reply tests/test_local_bindings.py:158]: If 7.3.2 is out before this gets merged, yes, let's remove it

Comment (YannickJadoul) [reply tests/test_local_bindings.py:158]: (unrelated change, btw, but something I bumped into, when testing locally)

Comment (henryiii) [reply tests/test_local_bindings.py:158]: I probably wouldn't drop it right away; approximately supporting 7.3.0+ would probably be ideal.

Comment (YannickJadoul) [reply tests/test_local_bindings.py:158]: Yes, that's probably best.

File: include/pybind11/detail/class.h [hunk 2]
```diff
@@ -24,6 +24,14 @@ PYBIND11_NAMESPACE_BEGIN(detail)
 #  define PYBIND11_SET_OLDPY_QUALNAME(obj, nameobj) setattr((PyObject *) obj, "__qualname__", nameobj)
 #endif
 
+inline std::string get_tp_name(PyTypeObject *type) {
+#if !defined(PYPY_VERSION)
```
Comment (henryiii) [inline include/pybind11/detail/class.h:28]: What happens in 7.3.2? Does this line start working?

Comment (YannickJadoul) [reply include/pybind11/detail/class.h:28]: Ugh .. I'd written this yesterday, but it was still a "pending review" ...

File: include/pybind11/detail/class.h [hunk 3]
```diff
@@ -24,6 +24,14 @@ PYBIND11_NAMESPACE_BEGIN(detail)
 #  define PYBIND11_SET_OLDPY_QUALNAME(obj, nameobj) setattr((PyObject *) obj, "__qualname__", nameobj)
 #endif
 
+inline std::string get_tp_name(PyTypeObject *type) {
```
Comment (rwgk) [inline include/pybind11/detail/class.h:27]: `get_fq_type_name` or even `get_fully_qualified_type_name`?
Somehow conveying that this function is meant to produce the fully qualified name seems important to me.

**Review by rwgk — COMMENTED** [review]
This looks like a great step in the right direction.
Are there existing unit tests for built-in types and nested classes? (If not it would be OK with me to add them later.)

File: include/pybind11/detail/class.h (hunk 3 above)
Comment (YannickJadoul) [reply include/pybind11/detail/class.h:27]: Yes, good point!

**Comment by YannickJadoul** [comment]
> Are there existing unit tests for built-in types and nested classes? (If not it would be OK with me to add them later.)

I think so, yes. Do the tests that I had to change (because PyPy is now consistent with CPython) count?

File: include/pybind11/detail/class.h (hunk 3 above)
Comment (YannickJadoul) [reply include/pybind11/detail/class.h]: Done! Thanks for the suggestion.

File: include/pybind11/detail/class.h [hunk 4]
```diff
@@ -476,9 +479,8 @@ extern "C" inline int pybind11_clear(PyObject *self) {
 inline void enable_dynamic_attributes(PyHeapTypeObject *heap_type) {
     auto type = &heap_type->ht_type;
 #if defined(PYPY_VERSION) && (PYPY_VERSION_NUM < 0x06000000)
-    pybind11_fail(std::string(type->tp_name) + ": dynamic attributes are "
-                                               "currently not supported in "
-                                               "conjunction with PyPy!");
+    pybind11_fail(get_tp_type(type) + ": dynamic attributes are currently not "
```
Comment (rwgk) [inline include/pybind11/detail/class.h:482]: This still shows `get_tp_type`. Oversight?

**Review by rwgk — APPROVED** [review]
Thanks Yannick! I think this is good to merge, to help you finishing up #2349. I didn't have a chance to look into existing unit test coverage beyond the changes in `test_class.py` here, but we can look into that separately, later, or as the need arises.

File: include/pybind11/detail/class.h (hunk 4 above)
Comment (YannickJadoul) [reply include/pybind11/detail/class.h:482]: Oh. That was even wrong :-| It used to be `get_tp_name`... This is part of the code is going away in #2456, though. That's why it's not tested.

**Comment by YannickJadoul** [comment]
OK, sounds good to me. Thanks, @rwgk!

**Review by henryiii — APPROVED** [review]

**Comment by henryiii** [comment]
This seems safe to put in; would like to slowly keep a little momentum heading into 2.6.0.

**Pull request merged**